server/src/main/java/org/elasticsearch/health/metadata/HealthMetadataService.java

```diff
@@ -124,39 +132,40 @@ private void enable(boolean enabled) {
             resetHealthMetadata("health-node-enabled");
         } else {
             clusterService.removeListener(clusterStateListener);
-            readyToPublish = false;
+            readyToPublishAsMaster = false;
         }
     }
 
     private void updateOnClusterStateChange(ClusterChangedEvent event) {
         final boolean wasMaster = event.previousState().nodes().isLocalNodeElectedMaster();
         isMaster = event.localNodeMaster();
         if (isMaster && wasMaster == false) {
-            readyToPublish = true;
+            readyToPublishAsMaster = true;
         } else if (isMaster == false) {
-            readyToPublish = false;
+            readyToPublishAsMaster = false;
         }
+
+        // Avoid scheduling updates to the taskQueue in case the cluster is recovering.
+        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
+            return;
+        }
+
         // Wait until every node in the cluster is upgraded to 8.5.0 or later
         if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
-            if (readyToPublish) {
+            if (readyToPublishAsMaster) {
                 resetHealthMetadata("health-metadata-update-master-election");
-                readyToPublish = false;
-            }
-        }
-    }
-
-    private void updateOnSettingsUpdated(UpsertHealthMetadataTask task) {
-        // We do not use the cluster state to check if this is the master node because the cluster state might not have been initialized
-        if (isMaster && enabled) {
-            ClusterState clusterState = clusterService.state();
-            if (clusterState.nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
-                taskQueue.submitTask("health-metadata-update", task, null);
+                readyToPublishAsMaster = false;
+            } else if (isMaster) {
+                var currentStoredHealthMetadata = HealthMetadata.getFromClusterState(event.state());
+                if (currentStoredHealthMetadata != null && this.localHealthMetadata.equals(currentStoredHealthMetadata) == false) {
```

Comment:
I don't want to bike shed too much here, thanks for iterating on this so far ❤️ 

I'm wondering here if we can simplify the logic here quite a bit (for e.g. it's a bit confusing to me that we have this condition here `if (currentStoredHealthMetadata != null` - as the `currentStoredHealthMetadata` could indeed be null - however, the `null` case it's handled on the previous branch of the if statement where we reset the health stuff)

What if we don't differentiate between a `reset` health metadata case and a regular `upsert` health metadata case (due to settings change) ? 
Also what if we drop the `readyToPublishAsMaster` flag? (it _seems_ that it just guards against the whole cluster being upgraded, however we have a designated `if` check for that - unless I'm missing something I think we could drop this flag altogether)

For e.g. this `updateOnClusterStateChange` method could look like so:

```
    private void updateOnClusterStateChange(ClusterChangedEvent event) {
        // Avoid scheduling updates to the taskQueue in case the cluster is recovering.
        if (event.state().blocks().hasGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK)) {
            return;
        }
        final boolean prevIsMaster = this.isMaster;
        if (prevIsMaster != event.localNodeMaster()) {
            this.isMaster = event.localNodeMaster();
        }

        if (this.isMaster) {
            // Wait until every node in the cluster is upgraded to 8.5.0 or later
            if (event.state().nodesIfRecovered().getMinNodeVersion().onOrAfter(Version.V_8_5_0)) {
                var currentStoredHealthMetadata = HealthMetadata.getFromClusterState(event.state());
                if (currentStoredHealthMetadata == null || this.localHealthMetadata.equals(currentStoredHealthMetadata) == false) {
                    logger.info("updating health to {}", this.localHealthMetadata);
                    taskQueue.submitTask("health-metadata-update", () -> this.localHealthMetadata, null);
                }
            }
        }
    }
```

What do you think?

Reply:
I've tested this locally and pushed a branch here in case it's helpful? https://github.com/andreidan/elasticsearch/tree/simplify-healthmetadataservice

Reply:
Indeed, since we check the health metadata from the cluster state we do not need the flag.

Reply:
damn, I didn't read you comment about the branch and wrote all this code by my own 🤦🏽